Move all lib functions out of main script

drpatelh opened the pull request:
I have already moved most of the bioler plate functions out of the main script into some appropriately named Groovy classes in the `lib/` folder. I have also renamed all functions to be `camelCase` for consistency. This PR removes the last remnants from the main script as inspired by https://github.com/nf-core/tools/issues/1043 so now the `main.nf` script has next to no biolerplate and is called via a single function that validates everything 🎉 

It would be awesome if others can double-check how I have named the files in `lib/` and also how I have organised the functions within these. This is quite important because if we get it right now then it means less issues with syncing downstream where filenames / function names etc have to be changed.

The current `lib/` folder I am planning to port to the DSL2 template can be found [here](https://github.com/drpatelh/rnaseq/tree/attributes/lib):
- [`Checks.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/Checks.groovy) - Standard checks to perform before the pipeline execution (not sure about this one? as some of the `params` are still nf-core specific)
- [`Completion.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/Completion.groovy) - Completion email and reporting class
- [`NfcoreSchema.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/NfcoreSchema.groovy) - nf-core Schema validation class
- [`Utils.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/Utils.groovy) - General utility functions 
- [`Workflow.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/Workflow.groovy) - All workflow specific validation and functions (if we have more than one workflow in the pipeline then this has holds the functions for all of them - do we separate them out into separate Groovy classes per workflow instead e.g. `lib/WorfklowIllumina.groovy`  and `lib/WorkflowNanopore.groovy`)

The current `main.nf` script I am planning to port to the DSL2 template can be found [here](https://github.com/drpatelh/rnaseq/blob/attributes/main.nf)

**SEE HERE FOR UPDATED STRUCTURE**: https://github.com/nf-core/rnaseq/pull/622#issuecomment-831599527

ewels commented on lib/Workflow.groovy:
```diff
@@ -15,12 +15,42 @@ class Workflow {
+        return summary_log
+    }
+
     public static void validateMainParams(workflow, params, json_schema, log) {
```
I'd be tempted to call this something else, as it does a lot more than just validate params now. `initialiseWorkflow` or `startYourEngines` or something? 😅

drpatelh replied:
Done https://github.com/drpatelh/rnaseq/blob/f1f5e891d4bcfd9c45965857b33b3cc66f2c2ce6/main.nf#L31

drpatelh replied:
Now `WorkflowMain.initialise(workflow, params, log)`

ewels commented on main.nf:
```diff
@@ -40,24 +25,18 @@ params.rsem_index   = Workflow.getGenomeAttribute(params, 'rsem')
+/* --          VALIDATE PARAMETERS             -- */
 ////////////////////////////////////////////////////
 
+def json_schema = "$projectDir/nextflow_schema.json"
```
Can we define this in the library as it's always the same? Could accept an optional function argument if we want to allow people to customise it..

drpatelh replied:
Wrote a function [here](https://github.com/drpatelh/rnaseq/blob/f1f5e891d4bcfd9c45965857b33b3cc66f2c2ce6/lib/NfcoreSchema.groovy#L16-L19) which does this. Tried various ways but `$projectDir` has to be accessed via evaluation with `workflow.projectDir` so this is the best way I could think of to do this.

ewels commented on main.nf:
```diff
@@ -40,24 +25,18 @@ params.rsem_index   = Workflow.getGenomeAttribute(params, 'rsem')
 /* --        NAMED PIPELINE WORKFLOW           -- */
 ////////////////////////////////////////////////////
 
+def summary_params = NfcoreSchema.paramsSummaryMap(workflow, params, json_schema)
```
This is for the email right? Maybe we can just call it directly where it's used. This is rather isolated now.

drpatelh replied:
Done [here](https://github.com/drpatelh/rnaseq/blob/f1f5e891d4bcfd9c45965857b33b3cc66f2c2ce6/workflows/rnaseq.nf#L11) and removed from `main.nf`

ewels reviewed (commented):
Looking good! Couple of minor comments..

KevinMenden commented:
Look reasonable to me! 

And I would vote for sticking to one `Workflow.groovy`  file for DSL2 pipeline irrespective of how many sub-workflows that pipeline has. Just nicer to have one common file for all pipelines in my opinion, and the functions inside could still be named appropriately. 

:+1:

ggabernet commented:
Looks good to me too, and I would also agree to all other suggestions, I love how clean `main.nf` stays then.

drpatelh commented:
Ok. I had a long bank holiday weekend to think about this :) 

I think I have now implemented something that organises the `lib/` functions in a way which makes it more obvious as to what each of the classes is doing. You can see the new structure [here](https://github.com/drpatelh/rnaseq/tree/attributes/lib).

I have deleted `Checks.groovy`, `Completion.groovy` and `Workflow.groovy` and copied across and organised their contents into the files below:

- [`NfcoreSchema.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/NfcoreSchema.groovy) - nf-core Schema validation class
- [`NfcoreTemplate.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/NfcoreTemplate.groovy) - nf-core template functions
- [`Utils.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/Utils.groovy) - General utility functions that can be used in any Nextflow pipeline
- [`WorkflowMain.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/WorkflowMain.groovy) - Validation and functions for the `main.nf` workflow
- [`WorkflowRnaseq.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/WorkflowRnaseq.groovy) - Validation and functions for the `workflows/rnaseq.nf` workflow
- [`WorkflowSraDownload.groovy`](https://github.com/drpatelh/rnaseq/tree/attributes/lib/WorkflowSraDownload.groovy) - Validation and functions for the `workflows/sra_download.nf` workflow

Now we have one file holding all nf-core related functions which is not only better for organisation but makes things more flexible for developers that want to spin-up and develop non-nf-core pipelines.

Separate `Workflow*.groovy` classes for all workflows in any given pipeline. Having spent quite a bit of time adding and amending functions when we had one class for all of these functions (previously `Workflow.groovy`) I realised that splitting things out would make it much easier to find functions specific to any given workflow, will make the template syncing easier and adheres to the modular ethos of DSL2.

Having also addressed @ewels comments we now have a stupendously minimal `main.nf` script 🥳

I am quite happy with the way this is looking now but as always feedback more than welcome so we can attempt to get it as right as possible given that this pipeline will be our latest working template for DSL2 when I get around to releasing it....

One thing I am not happy about is how we standardise Groovy comments in the pipeline template. Be nice to get some suggestions e.g. any better alternatives to [this](https://github.com/drpatelh/rnaseq/blob/f1f5e891d4bcfd9c45965857b33b3cc66f2c2ce6/main.nf#L14-L16)? I don't like the fact that things have to be manually centered...

KevinMenden commented:
Okay, looks good although I still prefer one `Workflow.groovy`. I'm not a fan of sticking everything into separate files whenever possible, as I don't think that this inevitably helps readability. And I imagine that often functionality will be common among the workflows. Then you either replicate the code or put it in another file again. I'm just a little concerned that all this DSL2 modularity gradually makes those pipelines very complex, and hard to understand for people other than the developers. 

But that's just my opinion. If you think it's better to stick it into separate files then go with it. Has its benefits, too :+1:

JoseEspinosa commented:
Wow looks wonderful :heart_eyes: 
Regarding the comments, I agree that is a bad idea to manually center them, personally, I'd go for something like:

```
/*
 * My first awesome comment
 */
```

Or otherwise, any other alternative that does not need to center them.

JoseEspinosa commented on lib/WorkflowRnaseq.groovy:
```diff
@@ -184,9 +150,7 @@ class Workflow {
-     * Get workflow summary for MultiQC
-     */
+    // Get workflow summary for MultiQC
     public static String paramsSummaryMultiqc(workflow, summary) {
```
Maybe there is something specific to the `rnaseq` pipeline but otherwise, could this function not be moved to `workflowMain.groovy`

drpatelh replied:
Yup, its specific to the rnaseq workflow in this pipeline which is why I put it here.

drpatelh commented:
Ok I think I am pretty much done here now. Be awesome if I can get a review and approval if everything is in order.

maxulysse commented on main.nf:
```diff
@@ -11,74 +11,58 @@
 
 nextflow.enable.dsl = 2
```
you forgot to left align `nf-core/rnaseq` on line 4

drpatelh replied:
Yup. I think this is how it is in the pipeline template [here](https://github.com/nf-core/tools/blob/dff9958cfb5b512abe62983ebf3574d8206bb433/nf_core/pipeline-template/main.nf#L4). But we can change it there too.

drpatelh replied:
How does this look? https://github.com/nf-core/rnaseq/pull/622/commits/db9d18f81460d31d6d30212fa4835fa932d957b0

drpatelh replied:
Added some appropriate links too whilst I was there :)

ewels commented on main.nf:
```diff
@@ -1,19 +1,19 @@
- Slack  : https://nfcore.slack.com/channels/rnaseq
+    Github : https://github.com/nf-core/rnaseq
+    Website: https://nf-co.re/rnaseq
+    lack  : https://nfcore.slack.com/channels/rnaseq
```
`Slack` not `lack` 🐛

ewels commented on main.nf:
```diff
@@ -1,84 +1,68 @@
- https://github.com/nf-core/rnaseq
+    Github : https://github.com/nf-core/rnaseq
+    Website: https://nf-co.re/rnaseq
+    lack  : https://nfcore.slack.com/channels/rnaseq
```
```suggestion
    Slack  : https://nfcore.slack.com/channels/rnaseq
```

drpatelh replied:
👀 Remnant of a late night indent exercise I suspect

drpatelh replied:
Fixed https://github.com/nf-core/rnaseq/pull/622/commits/349d083f95a0e995a615424f3f6671ef1e70ebba

JoseEspinosa commented on lib/NfcoreSchema.groovy:
```diff
@@ -12,17 +12,24 @@ import groovy.json.JsonSlurper
 import groovy.json.JsonBuilder
 
 class NfcoreSchema {
```
`NfCoreSchema`? and in the same way `NfCoreTemplate`... :roll_eyes:

KevinMenden replied:
I don't know that difference between "C" "c" can be hard to catch, I would rather leave it simple with "Nfcore"

drpatelh replied:
Yep.  I think we should just keep it the same as it is now. Still a bit annoyed that we have to use `N` to make Groovy happy :)

JoseEspinosa commented on lib/NfcoreTemplate.groovy:
```diff
@@ -1,8 +1,49 @@
+            // related: https://github.com/nextflow-io/nextflow/issues/813
+            assert params.outdir.startsWith('s3:')       : "Outdir not on S3 - specify S3 Bucket to run on AWSBatch!"
+            // Prevent trace files to be stored on S3 since S3 does not support rolling files.
+            assert !params.tracedir.startsWith('s3:')    :  "Specify a local tracedir or run without trace! S3 cannot be used for tracefiles."
```
Thought this was solved with `overwrite = true` as reported on [this](https://github.com/nextflow-io/nextflow/issues/813) nextflow closed issue.

drpatelh replied:
Yes, I believe you are right 👍🏽 Think we can remove that now because we have gone with using a unique filename https://github.com/nf-core/rnaseq/blob/9d05b22569fc8db0407f9e2bf03e60906048b327/nextflow.config#L204 as opposed to `overwrite = true`.

drpatelh replied:
Excellent spot! https://github.com/nf-core/rnaseq/pull/622/commits/4bf388848554b54de0fa3ea77eedea119b15d3e5

JoseEspinosa commented on lib/WorkflowSraDownload.groovy:
```diff
@@ -0,0 +1,19 @@
+     * Print a warning after SRA download has completed
+     */
+    public static void sraDownloadWarn(log) {
+        log.warn "=============================================================================\n" +
```
Might be worth mentioning here the strandedness, I know is nicely explained in the `usage.md` ("public databases don't reliably hold information such as strandedness information so you may need to amend these entries too") so up to you...

drpatelh replied:
I did have that info in originally but I removed it to make it more generic so it will be easier to copy and paste this across when we add an independent workflow for this. For example, the strandedness stuff isn't applicable to viralrecon where we are using exactly the same workflow. Added back in anyway https://github.com/nf-core/rnaseq/pull/622/commits/41dbf0ec3fd96550c728de94cc7bcf6d67d9c3c9

JoseEspinosa reviewed (approved):
Beautiful! Just a couple of comments that you may decide not to change :hear_no_evil:
